File: saveit-backend/api/info.py
import json
import re
from http.server import BaseHTTPRequestHandler
import yt_dlp


def detect_platform(url):
    patterns = {
        "youtube": r"(youtube\.com|youtu\.be)",
        "tiktok": r"tiktok\.com",
        "instagram": r"instagram\.com",
        "facebook": r"(facebook\.com|fb\.watch|fb\.com)",
        "twitter": r"(twitter\.com|x\.com)",
    }
    for platform, pattern in patterns.items():
        if re.search(pattern, url, re.IGNORECASE):
            return platform
    return "unknown"


def get_video_info(url):
    platform = detect_platform(url)

    ydl_opts = {
        "quiet": True,
        "no_warnings": True,
        "extract_flat": False,
        "skip_download": True,
        "noplaylist": True,
    }

    # TikTok: no watermark trick
    if platform == "tiktok":
        ydl_opts["extractor_args"] = {
            "tiktok": {"api_hostname": "api22-normal-c-useast2a.tiktokv.com"}
        }

    with yt_dlp.YoutubeDL(ydl_opts) as ydl:
        info = ydl.extract_info(url, download=False)

    if not info:
        raise ValueError("Tidak bisa mengambil info video")

    # Build format list
    formats = []
    seen = set()

    raw_formats = info.get("formats") or []
    for f in raw_formats:
        vcodec = f.get("vcodec", "none")
        acodec = f.get("acodec", "none")
        ext = f.get("ext", "")
        height = f.get("height")
        filesize = f.get("filesize") or f.get("filesize_approx")

        # Only video+audio combined, or video-only (we'll note it)
        if vcodec == "none" or ext not in ("mp4", "webm", "mov"):
            continue

        quality_label = f"{height}p" if height else "Unknown"
        key = (quality_label, ext)
        if key in seen:
            continue
        seen.add(key)

        formats.append({
            "format_id": f.get("format_id"),
            "url": f.get("url"),
            "ext": ext,
            "quality": quality_label,
            "height": height or 0,
            "filesize": filesize,
            "has_audio": acodec != "none",
        })

    # Sort by quality descending
    formats.sort(key=lambda x: x["height"], reverse=True)

    # Fallback: if no formats found, use direct url
    if not formats and info.get("url"):
        formats.append({
            "format_id": "best",
            "url": info["url"],
            "ext": info.get("ext", "mp4"),
            "quality": "Best",
            "height": 0,
            "filesize": None,
            "has_audio": True,
        })

    thumbnail = info.get("thumbnail") or ""
    # Use the best thumbnail
    thumbnails = info.get("thumbnails") or []
    if thumbnails:
        thumbnail = thumbnails[-1].get("url", thumbnail)

    return {
        "platform": platform,
        "title": info.get("title", "Video"),
        "thumbnail": thumbnail,
        "duration": info.get("duration"),
        "uploader": info.get("uploader") or info.get("channel", ""),
        "view_count": info.get("view_count"),
        "formats": formats[:8],  # max 8 format options
    }


class handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.end_headers()

    def do_POST(self):
        try:
            length = int(self.headers.get("Content-Length", 0))
            body = self.rfile.read(length)
            data = json.loads(body)
            url = data.get("url", "").strip()

            if not url:
                raise ValueError("URL tidak boleh kosong")

            result = get_video_info(url)

            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.send_header("Access-Control-Allow-Origin", "*")
            self.end_headers()
            self.wfile.write(json.dumps({"success": True, "data": result}).encode())

        except Exception as e:
            self.send_response(400)
            self.send_header("Content-Type", "application/json")
            self.send_header("Access-Control-Allow-Origin", "*")
            self.end_headers()
            self.wfile.write(
                json.dumps({"success": False, "error": str(e)}).encode()
            )

    def log_message(self, format, *args):
        pass

File: api/info.js
const COBALT_INSTANCES = [
  'https://api.cobalt.tools',
  'https://cobalt.ggtyler.dev',
  'https://cobalt-api.kwiatekmiki.com',
];

function detectPlatform(url) {
  if (/youtube\.com|youtu\.be/.test(url)) return 'youtube';
  if (/tiktok\.com/.test(url)) return 'tiktok';
  if (/instagram\.com/.test(url)) return 'instagram';
  if (/facebook\.com|fb\.watch/.test(url)) return 'facebook';
  if (/twitter\.com|x\.com/.test(url)) return 'twitter';
  return 'unknown';
}

async function tryCobalt(url, instance) {
  const res = await fetch(`${instance}/`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    },
    body: JSON.stringify({
      url,
      videoQuality: '1080',
      audioFormat: 'mp3',
    }),
    signal: AbortSignal.timeout(10000),
  });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  return await res.json();
}

async function getVideoInfo(url) {
  const platform = detectPlatform(url);
  let lastError = null;

  for (const instance of COBALT_INSTANCES) {
    try {
      const data = await tryCobalt(url, instance);
      if (data.status === 'error') throw new Error(data.error?.code || 'cobalt error');

      const formats = [];

      if (data.status === 'picker') {
        data.picker.forEach((item, i) => {
          formats.push({
            format_id: `item_${i}`,
            url: item.url,
            ext: 'mp4',
            quality: item.type === 'photo' ? `Foto ${i+1}` : `Video ${i+1}`,
            has_audio: item.type !== 'photo',
          });
        });
      } else if (data.url) {
        formats.push({
          format_id: 'best',
          url: data.url,
          ext: 'mp4',
          quality: 'Best Quality',
          has_audio: true,
        });
      }

      // oEmbed untuk metadata
      let title = platform.charAt(0).toUpperCase() + platform.slice(1) + ' Video';
      let thumbnail = '';
      let uploader = '';

      const oembedMap = {
        youtube: `https://www.youtube.com/oembed?url=${encodeURIComponent(url)}&format=json`,
        tiktok: `https://www.tiktok.com/oembed?url=${encodeURIComponent(url)}`,
        twitter: `https://publish.twitter.com/oembed?url=${encodeURIComponent(url)}`,
      };

      if (oembedMap[platform]) {
        try {
          const oe = await fetch(oembedMap[platform], { signal: AbortSignal.timeout(4000) });
          if (oe.ok) {
            const oed = await oe.json();
            title = oed.title || title;
            thumbnail = oed.thumbnail_url || thumbnail;
            uploader = oed.author_name || uploader;
          }
        } catch {}
      }

      return { platform, title, thumbnail, uploader, formats };
    } catch (err) {
      lastError = err;
    }
  }
  throw new Error(`Gagal: ${lastError?.message}`);
}

export default async function handler(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') return res.status(200).end();

  if (req.method === 'GET') {
    return res.status(200).json({ status: 'SaveIt API is running!' });
  }

  if (req.method !== 'POST') {
    return res.status(405).json({ success: false, error: 'Method not allowed' });
  }

  try {
    const { url } = req.body;
    if (!url?.trim()) return res.status(400).json({ success: false, error: 'URL kosong' });

    const data = await getVideoInfo(url.trim());
    return res.status(200).json({ success: true, data });
  } catch (err) {
    return res.status(400).json({ success: false, error: err.message });
  }
}
